File: node/test/properties/expansionOrder.property.test.ts
import { describe, it, expect } from "vitest";
import { pool } from "../../src/db/pool.js";
import { runExpansionCycle } from "../../src/worker/expansion.js";
import { createTenant, createEndpoint } from "../fixtures.js";
import { mulberry32, getTestSeed, randomInt } from "./prng.js";

async function publishEvent(tenantId: string, idempotencyKey: string): Promise<string> {
  const { rows } = await pool.query<{ id: string }>(
    `INSERT INTO events (tenant_id, idempotency_key, type, payload) VALUES ($1, $2, 'order.created', '{}') RETURNING id`,
    [tenantId, idempotencyKey],
  );
  return rows[0]!.id;
}

// Drains every pending_expansion event by racing several concurrent workers
// against the real advisory-lock claim — the actual concurrency-safety
// property under test, not simulated.
async function drainExpansion(concurrency: number): Promise<void> {
  for (;;) {
    const results = await Promise.all(Array.from({ length: concurrency }, () => runExpansionCycle(pool)));
    if (results.every((didWork) => !didWork)) return;
  }
}

// REVIEW.md F-1 / docs/adr/0001: per-endpoint delivery order must equal
// events.seq order, even when expansion runs fully concurrently — the
// per-tenant advisory lock exists specifically to make this true under
// concurrent workers, not just under the sequential expansion the earlier
// example-based tests (expansion.test.ts) exercise.
describe("expansion order property: per-endpoint delivery order equals events.seq order", () => {
  it("holds for a randomized number of events published to one endpoint, expanded by racing concurrent workers", async () => {
    const seed = getTestSeed("expansion-order");
    const rng = mulberry32(seed);
    const eventCount = randomInt(rng, 5, 15);
    const concurrency = randomInt(rng, 2, 5);

    const { id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);

    const publishedEventIdsInOrder: string[] = [];
    for (let i = 0; i < eventCount; i++) {
      publishedEventIdsInOrder.push(await publishEvent(tenantId, `expansion-order-${seed}-${i}`));
    }

    await drainExpansion(concurrency);

    const { rows: deliveries } = await pool.query<{ event_id: string }>(
      "SELECT event_id FROM deliveries WHERE endpoint_id = $1 ORDER BY seq",
      [endpoint.id],
    );

    expect(deliveries.map((d) => d.event_id)).toEqual(publishedEventIdsInOrder);
  });
});

File: go/internal/worker/expansion_order_property_test.go
package worker_test

import (
	"context"
	"math/rand"
	"sync"
	"testing"

	"github.com/stretchr/testify/require"

	"webhooks-go/internal/properties"
	"webhooks-go/internal/testsupport"
	"webhooks-go/internal/worker"
)

// REVIEW.md F-1 / docs/adr/0001: per-endpoint delivery order must equal
// events.seq order, even when expansion runs fully concurrently — the
// per-tenant advisory lock exists specifically to make this true under
// concurrent workers, not just under the sequential expansion the earlier
// example-based tests (expansion_test.go) exercise.
func TestExpansionOrderPropertyHoldsUnderConcurrentWorkers(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ctx := context.Background()

	seed := properties.GetTestSeed("expansion-order")
	rng := rand.New(rand.NewSource(seed))
	eventCount := 5 + rng.Intn(11) // [5, 15]
	concurrency := 2 + rng.Intn(4) // [2, 5]

	tenantID, _ := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})

	publishedEventIDs := make([]string, eventCount)
	for i := 0; i < eventCount; i++ {
		publishedEventIDs[i] = testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{Status: "pending_expansion"})
	}

	// Drains every pending_expansion event by racing several concurrent
	// workers against the real advisory-lock claim — the actual
	// concurrency-safety property under test, not simulated.
	for {
		var wg sync.WaitGroup
		results := make([]bool, concurrency)
		for i := 0; i < concurrency; i++ {
			wg.Add(1)
			go func(i int) {
				defer wg.Done()
				did, err := worker.RunExpansionCycle(ctx, pool)
				require.NoError(t, err)
				results[i] = did
			}(i)
		}
		wg.Wait()
		anyWork := false
		for _, did := range results {
			anyWork = anyWork || did
		}
		if !anyWork {
			break
		}
	}

	rows, err := pool.Query(ctx, "SELECT event_id FROM deliveries WHERE endpoint_id = $1 ORDER BY seq", endpointID)
	require.NoError(t, err)
	defer rows.Close()
	var deliveredEventIDs []string
	for rows.Next() {
		var id string
		require.NoError(t, rows.Scan(&id))
		deliveredEventIDs = append(deliveredEventIDs, id)
	}
	require.NoError(t, rows.Err())
	require.Equal(t, publishedEventIDs, deliveredEventIDs)
}
